### PR TITLE
[feature-wip](executor)Add cgroup cpu controller

### PR DESCRIPTION
**Pull request by wangbo**
## Proposed changes

First pr of #24051

1 Add cgroup cpu controller.
2 support cpu hard limit for all queries.
3 Add ```cpu_hard_limit``` for workload group.

### CODE REVIEW

**Comment by HappenLee** [comment]
run buildall

File: be/src/agent/cgroup_cpu_ctl.h [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+namespace doris {
+
+class CgroupCpuCtl {
+public:
+    CgroupCpuCtl() {}
+    virtual ~CgroupCpuCtl() {}
+
+    virtual Status init();
+
+    virtual Status modify_cg_cpu_hard_limit_no_lock(int cpu_hard_limit) { return Status::OK(); }
```
Comment (HappenLee) [inline be/src/agent/cgroup_cpu_ctl.h:38]: why not be pure virtual function

File: be/src/agent/cgroup_cpu_ctl.cpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+void CgroupCpuCtl::update_cpu_hard_limit(int cpu_hard_limit) {
+    if (!_init_succ) {
+        return;
+    }
+    std::lock_guard<std::shared_mutex> w_lock(_lock_mutex);
+    if (_cpu_hard_limit != cpu_hard_limit) {
+        _cpu_hard_limit = cpu_hard_limit;
+        modify_cg_cpu_hard_limit_no_lock(cpu_hard_limit);
```
Comment (HappenLee) [inline be/src/agent/cgroup_cpu_ctl.cpp:42]: here need to process return the Status of `modify_cg_cpu_hard_limit_no_lock`

File: be/src/runtime/exec_env_init.cpp [hunk 3]
```diff
@@ -226,15 +226,26 @@ Status ExecEnv::init_pipeline_task_scheduler() {
         executors_size = CpuInfo::num_cores();
     }
 
+    if (config::enable_cgroup_cpu_controller) {
+        _cgroup_cpu_ctl = new CgroupV1CpuCtl();
```
Comment (HappenLee) [inline be/src/runtime/exec_env_init.cpp:230]: need safe delete to prevent the mem leak

File: be/src/agent/cgroup_cpu_ctl.cpp (hunk 2 above)
Comment (wangbo) [reply be/src/agent/cgroup_cpu_ctl.cpp:42]: cpu hard limit should be modify when a query arrives with workload group.
So if modify cgroup failed, query should not be blocked or failed.
This exception should be  handled when user find workload group may not work and see be.INFO

**Comment by wangbo** [comment]
run buildall

**Comment by wangbo** [comment]
run buildall

**Comment by wangbo** [comment]
run buildall

File: be/src/common/config.cpp [hunk 4]
```diff
@@ -1088,6 +1088,11 @@ DEFINE_Int16(bitmap_serialize_version, "1");
 // the count of thread to group commit insert
 DEFINE_Int32(group_commit_insert_threads, "10");
 
+// cgroup
+DEFINE_Bool(enable_cgroup_cpu_controller, "false");
```
Comment (yiguolei) [inline be/src/common/config.cpp:1092]: keep only one config: doris_cgroup_path

File: be/src/runtime/exec_env_init.cpp [hunk 5]
```diff
@@ -228,17 +228,27 @@ Status ExecEnv::init_pipeline_task_scheduler() {
         executors_size = CpuInfo::num_cores();
     }
 
+    if (config::enable_cgroup_cpu_controller) {
```
Comment (yiguolei) [inline be/src/runtime/exec_env_init.cpp:231]: just check doris_cgroup_path != empty and it is exist

File: be/src/runtime/exec_env_init.cpp [hunk 6]
```diff
@@ -228,17 +228,28 @@ Status ExecEnv::init_pipeline_task_scheduler() {
         executors_size = CpuInfo::num_cores();
     }
 
+    std::string cgroup_cu_path = config::doris_cgroup_cpu_path;
+    if (!cgroup_cu_path.empty()) {
```
Comment (HappenLee) [inline be/src/runtime/exec_env_init.cpp:232]: why here need a copy ？
```
if ( config::doris_cgroup_cpu_path.empty())
```

**Comment by wangbo** [comment]
run buildall

**Comment by wangbo** [comment]
run buildall

**Comment by wangbo** [comment]
run buildall

**Comment by wangbo** [comment]
run buildall

**Comment by wangbo** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by zclllyybb — APPROVED** [review]

**Pull request merged**